source/slang/slang-capability.cpp

```diff
@@ -170,121 +158,107 @@ bool isCapabilityDerivedFrom(CapabilityAtom atom, CapabilityAtom base)
     const auto& info = kCapabilityNameInfos[Index(atom)];
     for (auto cur : info.canonicalRepresentation)
     {
-        for (auto cbase : cur)
-            if (asAtom(cbase) == base)
-                return true;
+        if (cur->contains((UInt)base))
+            return true;
     }
 
     return false;
 }
 
 //// CapabiltySet
 
-void CapabilitySet::addToTargetCapabilityWithValidUIntSetAndTargetAndStage(CapabilityName target, CapabilityName stage, CapabilityAtomSet setToAdd)
+// MSVC incorrectly throws warning
+#if defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable:4702)
+#endif
+CapabilityAtom getTargetAtomInSet(const CapabilityAtomSet& atomSet)
 {
-    SLANG_ASSERT(target != CapabilityName::Invalid && stage != CapabilityName::Invalid);
-    auto stageAtom = asAtom(stage);
-    auto targetAtom = asAtom(target);
-    CapabilityTargetSet& targetSet = m_targetSets[targetAtom];
-    targetSet.target = targetAtom;
-    targetSet.shaderStageSets.reserve(kCapabilityStageCount);
-
-    auto& localStageSets = targetSet.shaderStageSets[stageAtom];
-    localStageSets.stage = stageAtom;
-
-    localStageSets.addNewSet(std::move(setToAdd));
+    auto targetSet = getAtomSetOfTargets();
+    CapabilityAtomSet out;
+    CapabilityAtomSet::calcIntersection(out, targetSet, atomSet);
+    for (auto i : out)
+        return (CapabilityAtom)i;
+    return CapabilityAtom::Invalid;
 }
 
-void CapabilitySet::addToTargetCapabilityWithTargetAndStageAtom(CapabilityName target, CapabilityName stage, const ArrayView<CapabilityName>& canonicalRepresentation)
+CapabilityAtom getStageAtomInSet(const CapabilityAtomSet& atomSet)
 {
-    // If no provided 'stage', set the capability as a target of all stages
-    if (stage == CapabilityName::Invalid)
-    {
-        auto info = _getInfo(CapabilityName::any_stage);
-        List<CapabilityName> newArr;
-        auto count = canonicalRepresentation.getCount();
-        newArr.setCount(count + 1);
-        memcpy(newArr.getBuffer(), canonicalRepresentation.getBuffer(), count * sizeof(CapabilityName));
-        m_targetSets[asAtom(target)].shaderStageSets.reserve(info.canonicalRepresentation.getCount());
-        for (auto i : info.canonicalRepresentation)
-        {
-            newArr[count] = i[0];
-            addToTargetCapabilityWithTargetAndStageAtom(target, i[0], newArr.getArrayView());
-        }
-        return;
-    }
-    
-    CapabilityAtomSet setToAdd = CapabilityAtomSet((UInt)CapabilityAtom::Count);
-    for(auto i : canonicalRepresentation)
-        setToAdd.add(asAtomUInt(i));
-
-    addToTargetCapabilityWithValidUIntSetAndTargetAndStage(target, stage, setToAdd);
+    auto stageSet = getAtomSetOfStages();
+    CapabilityAtomSet out;
+    CapabilityAtomSet::calcIntersection(out, stageSet, atomSet);
+    for(auto i : out)
+        return (CapabilityAtom)i;
+    return CapabilityAtom::Invalid;
```

Comment:
This looks little strange.
It will return either Invalid or the first element of `out`.
I don't see why `for` loop is used.
That might be why you were getting the unreachable code warning

Reply:
A `for` loop is used since `out` is a UIntSet. A UIntSet is non trivial to get element's from and check it empty. Due to this, a iterator is used to check for a first element to keep the code readable.

The `for` loop is why MSVC misdiagnosis the code. This is just a MSVC compiler bug.

Reply:
I wouldn't say it is a bug because it is just warning to prevent common mistakes.
It is to warn that there "might" be a mistake.

Reply:
I recommend you not to silence the warning and change to code as following,
```
    auto iter = out.begin();
    if (iter == out.end())
        return CapabilityAtom::Invalid;
    return (CapabilityAtom )*iter;
```